Fix certificate revocation

buchdag opened the pull request:
As stated in #250 the way CRL is currently handled mean that it will require a container restart (contrary to what is stated in [clients.md](https://github.com/kylemanna/docker-openvpn/blob/master/docs/clients.md) ) to pick up the changes and that revocations past the first one won't work at all.

This pull request attempt to fix this by : 

- having `ovpn_initpki` generate a CRL so that `--crl-verify /etc/openvpn/crl.pem` will be passed to OpenVPN even on the container's first launch, removing the need for a container restart.
- relying on a copy of **easyrsa**'s **crl.pem** rather than a hardlink that'll get broken every time `easy-rsa gen-crl` is used , and replacing that copy with a newer crl if it is present on container launch.
- revoking clients using a new script that handle the certificate revocation, CRL generation, and creation of a new copy of **crl.pem** with proper read permissions for OpenVPN.

Plus updates of the relevant documentation files.

kylemanna commented:
Thanks for taking the time to work on this.  Many people have had issues with this (see issue #101), but nobody has taken the time to submit a PR! After reviwing your patch I think you've identified the problem and look to have solved it.

One request: Can you review the unit tests run by Jenkins to make sure this doesn't regress?  Ideally, it would fail with the current implementation and pass with your implementation.

Thanks!

buchdag commented:
@kylemanna I started poking around Travis CI and the existing tests, I'll come back to you as soon as I'll have a working unit test to merge.

buchdag commented:
I'm making progress but there is something I cannot figure : how do you get to echo something to Travis CI's console during test ? I thought that echoing to stderr was the way to go but `>&2 echo` does not work.

Travis CI is supposed to collect both stdout and stderr, yet most of the tests here are silent save for command errors like `/client/wait-for-connect.sh: connect: Connection refused`

Am I missing something somewhere else ? In [docker-openvpn/test/run.sh](https://github.com/kylemanna/docker-openvpn/blob/master/test/run.sh) maybe ?

kylemanna commented:
The `stdout` is consume by the `test.sh` script so that it can be used to pass/fail tests based on stdout strings compared to `expected-std-out.txt` file if present.

The way I had always debugged scripts like `test/tests/basic/run.sh` was to enable shell command execution tracing via `stderr`:

    DEBUG=1 test/run.sh -t basic kylemanna/openvpn

Expected output:

	$ DEBUG=1 test/run.sh -t basic kylemanna/openvpn
	testing kylemanna/openvpn
			'basic' [1/1]...+ OVPN_DATA=basic-data
	+ CLIENT=travis-client
	+ IMG=kylemanna/openvpn
	+++ dirname /home/nitro/tmp/docker-openvpn/test/tests/basic/run.sh
	++ readlink -f /home/nitro/tmp/docker-openvpn/test/tests/basic/../../client
	+ CLIENT_DIR=/home/nitro/tmp/docker-openvpn/test/client
	+ ip addr ls
	++ ip -4 -o addr show scope global
	++ awk '{print $4}'
	++ head -n1
	++ sed -e 's:/.*::'
	+ SERV_IP=192.168.10.10
	+ docker run -v basic-data:/etc/openvpn --rm kylemanna/openvpn ovpn_genconfig -u udp://192.168.10.10
	+ docker run -v basic-data:/etc/openvpn --rm -it -e EASYRSA_BATCH=1 -e 'EASYRSA_REQ_CN=Travis-CI Test CA' kylemanna/openvpn ovpn_initpki nopass
	+ docker run -v basic-data:/etc/openvpn --rm -it kylemanna/openvpn easyrsa build-client-full travis-client nopass
	+ docker run -v basic-data:/etc/openvpn --rm kylemanna/openvpn ovpn_getclient travis-client
	+ tee /home/nitro/tmp/docker-openvpn/test/client/config.ovpn
	+ docker run -v basic-data:/etc/openvpn --rm kylemanna/openvpn ovpn_listclients
	+ grep travis-client
	+ sudo iptables -N DOCKER
	iptables: Chain already exists.
	+ echo 'Firewall already configured'
	+ sudo iptables -I FORWARD -j DOCKER
	+ docker run --rm --net=host --privileged --volume /home/nitro/tmp/docker-openvpn/test/client:/client kylemanna/openvpn /client/wait-for-connect.sh
	+ docker run --name ovpn-test -v basic-data:/etc/openvpn --rm -p 1194:1194/udp --privileged kylemanna/openvpn
	/client/wait-for-connect.sh: connect: Connection refused
	/client/wait-for-connect.sh: line 19: /dev/tcp/127.0.0.1/9999: Connection refused
	iptables: No chain/target/match by that name.
	RTNETLINK answers: File exists
	+ kill %1
	+ cat
	RTNETLINK answers: Operation not permitted
	RTNETLINK answers: Operation not permitted
	RTNETLINK answers: Operation not permitted
	passed

Alternatively you could patch modify `test/run.sh` while you work on the tests, but the output will be pretty much garbage with screwed up white space:

```diff --git a/test/run.sh b/test/run.sh
index 47655fe..fdb40eb 100755
--- a/test/run.sh
+++ b/test/run.sh
@@ -171,6 +171,7 @@ for dockerImage in "$@"; do
                        if [ -x "$script" -a ! -d "$script" ]; then
                                # TODO dryRun logic
                                if output="$("$script" $dockerImage)"; then
+                                       echo $output
                                        if [ -f "$scriptDir/expected-std-out.txt" ] && ! d="$(echo "$output" | diff -u "$scriptDir/expected-std-out.txt" - 2>/dev/null)"; then
                                                echo 'failed; unexpected output:'
                                                echo "$d"
```


I haven't looked at the upstream docker testing repository where this stuff came from, but maybe it has improved and we could leverage it to improve the current build.

buchdag commented:
@kylemanna > should I rather include error messages that detail why the test failed like this : 

```
if sudo test ! -f /var/lib/docker/volumes/$OVPN_DATA/_data/pki/crl.pem
    then >&2 echo "CRL not generated during pki initialization."
    exit 2
fi
```

or skip the error messages, write something shorter like : 

`sudo test -f /var/lib/docker/volumes/$OVPN_DATA/_data/pki/crl.pem`

and let `set -e` handle the script failure ?

buchdag commented:
So far my test is doing the following : 

1. check if a `crl.pem` file is present in `/etc/openvpn/pki/` after running `ovpn_initpki`.
2. after container startup, check if this `crl.pem` file has been copied to `/etc/openvpn/crl.pem`  and that the copy is readable by openvpn.
3. generate a client certificate / configuration.
4. test client connection with `wait-for-connect.sh`.
5. revoke the client certificate (with key, crt an req deletion) using `ovpn_revokeclient`.
6. check that it does not show up in `ovpn_listclients`.
7. test connection again and fail if `wait-for-connect.sh` succeeds.
8. generate a new client cert / conf with the same CN.
9. goes through 4 to 7 again for the new cert.

By the way `ovpn_listclients` use the `crl.pem` that is in `/etc/openvpn/pki/`, not the one that the openvpn server will use, should it be modified to use `/etc/openvpn/crl.pem` instead ?

buchdag commented:
Regarding the last commit to ovpn_run : I thought that rather than having the revocation script make the necessary copy of the the new `/etc/openvpn/pki/crm.pem` to `/etc/openvpn/crl.pem`, it might be better to have `ovpn_run` execute a short function in the background that check every seconds (or .1 seconds maybe) if `/etc/openvpn/crl.pem` is newer than `/etc/openvpn/pki/crm.pem` and if so copy the former to the latter.

This way "manual" revocation using `easyrsa revoke your_client` then `easyrsa gen-crl` are also automatically picked up by the running container.

Does that sound like a good idea or should I revert to the `ovpn_revokeclient` script making the copy itself and `ovpn_run` not running a background function ?

kylemanna commented:
I'd try to keep the test as simple as possible because testing networking containers on Travis-CI is already convoluted.

As long as you can test that clients are no longer able to connect to the server after revoked and the server is restarted (several iterations of this of course) as documented, then I think the majority of the test is acceptable.

I'd break down some of your tests in to completely different tests if they don't depend on the server operation. For example, "6. check that it does not show up in ovpn_listclients" is second to the server flat out rejecting revoked certificates.  This could be skipped now, or added as another test that checks sanity things like this.

I favor more small tests over one monolithic hard to understand test.  It also makes it easier to troubleshoot small bite size tests when they break.

> Regarding the last commit to ovpn_run : I thought that rather than having the revocation script make the necessary copy of the the new /etc/openvpn/pki/crm.pem to /etc/openvpn/crl.pem, it might be better to have ovpn_run execute a short function in the background that check every seconds (or .1 seconds maybe) if /etc/openvpn/crl.pem is newer than /etc/openvpn/pki/crm.pem and if so copy the former to the latter.

This is complicated for a few reasons.  First, the bash shell goes away on the last line of `ovpn_run` when `exec` is called.  At that point `ovpn_run` is loaded in the process in place of `bash` and takes over, bash is no more.  This enforces a typical convention of Docker containers where there is only one running process in the container.

Alternatively, something else could be forked to try and copy the script, but that sounds overly complicated. The goal for this Docker image is to cater to the 90% use case where people want a quick + simple OpenVPN container.  This repository has become pretty popular and all the while only a handful of people have ever noticed the CRL bug that your pull request aims to fix. 

Really, the monitoring of the CRL should be fixed in OpenVPN itself (or maybe the real bug is that the hardlink breaks when the CRL is re-generated by easyrsa, perhaps we could fix this in the first place and OpenVPN would keep the file open and update itself?).  Perhaps the OpenVPN server should keep the file descriptor for the CRL open after dropping privileges to the un-privileged user so that it could be notified of changes using something like the Linux `inotify` API at which point it could re-read it.

All that said, I think restarting the server is the simplest compromise for now.

kylemanna commented:
Actually here is the real problem with the hard link breaking:

https://github.com/OpenVPN/easy-rsa/blob/5a429d22c78604c95813b457a8bea565a39793fa/easyrsa3/easyrsa#L742-L758

The problem is that easyrsa creates a temporary file, writes the CRL to that file, and the moves the CRL over the top of the previous one.  Presumably to make it an atomic operation to avoid corruption in a power failure (my understanding is that `mv` aka `rename()` syscall is typically atomic on journaling filesystems) or system crash at the exact same time.

What we could do instead if rewrite to the same file, that way the hardlink I originally used would work as expected and the permission would be preserved and openvpn would still have the CRL file open I think.

buchdag commented:
I agree on the reasons why `ovpn_run` should not spawn another process beside the main one, I reverted that commit.

> As long as you can test that clients are no longer able to connect to the server after revoked and the server is restarted (several iterations of this of course) as documented, then I think the majority of the test is acceptable.

What I have now in the test script is : 

- initialize config and pki then start the server
- generate a first client certificate, revoke it, test for failed connection
- generate a second client certificate (different CN), revoke it, test for failed connection
- restart the server, test for failed connection with the second client

Test fails with `kylemanna:master` branch, pass with `buchdag:buchdag-revoke1`.

It's not really testing for a refused connection, just for a failed `wait-for-connect.sh`. I'm not sure if this approach is enough or if something like a `wait-for-disconnect.sh` should be created instead.

Got rid of the rest for now.

kylemanna commented:
I think the test you described is good enough for now if it fails on my master and works on your branch.  With a few clean-ups I think we're ready to merge:

I was looking at your Travis-CI branch and had a few comments:
1. Please squash the commits in to only a few when you get to that point as opposed to a bunch of commits that add and delete things that were just added.  Break them in to smallest atomic units, i.e. first commit 
 could be add the test that detects the breakage, the second commit could be the solution to the breakage.  Shorter git logs are easier to follow.
2. Minor style nit pick with the `if` statements in the test, you have:
    ```
    if ... 
        then >&2 echo "Client was able to connect after revocation."
        exit 2
   fi
   ```
    to a more common convention that matches other code in the repo as well more elaborate failure message:
    ```
    if ... ; then
        echo "Client was able to connect after revocation test #2." >&2
        exit 2
   fi
   ```

Thanks for the work!

buchdag commented:
Commits squashed, style corrected and failure message elaborated.

I moved the client revocation test commit to the top when I rebased, I don't know why it stood at the bottom :|

kylemanna commented:
Looks great, Travis-CI is running the test and says it passes.  Good enough for me.

https://travis-ci.org/kylemanna/docker-openvpn/builds/230826694#L605-L610

rahulwa commented:
I believe that its important fix, so new version should be tagged.
